ingestion-beam/src/main/java/com/mozilla/telemetry/Decoder.java

```diff
@@ -72,11 +75,22 @@ public static PipelineResult run(DecoderOptions.Parsed options) {
             .apply("ParseUri", ParseUri.of()).failuresTo(failureCollections))
 
         // Special case: decryption of Pioneer payloads
-        .map(p -> options.getPioneerEnabled() ? p //
-            .apply(DecryptPioneerPayloads.of(options.getPioneerMetadataLocation(),
-                options.getPioneerKmsEnabled(), options.getPioneerDecompressPayload())) //
-            .failuresTo(failureCollections) //
-            : p)
+        // Both legacy messages sent to the `telemetry.pioneer-study` doctype
+        // and the glean.js-styled messages are supported respectively
+        .map(p -> options.getPioneerEnabled() ? PCollectionList.of(p //
+            .apply("FilterPioneer", Filter.by(new DecryptPayloads.PioneerPredicate())) //
```

Comment:
I initially tried going down this route which was the original reason for the DecryptPayloads class. I had trouble composing several `WithFailures` transforms so the wrapping transform also satisfies `WithFailures` where the source of `failuresTo` is the flat collection of the pioneer/rally errors. This is why the logic got moved up into the main `Decoder` class where the errors collection is in scope.

Reply:
We had this same problem for AET, if I'm understanding correctly: https://github.com/mozilla/gcp-ingestion/blob/c05fbabcb44c5a8a290be311a87951728cf587b6/ingestion-beam/src/main/java/com/mozilla/telemetry/aet/AetProcessing.java

Within AetProcessing, we create a fresh failureCollections array, and then we flatten it before returning an explicit `WithFailures.Result.of(output, errors)`.

Reply:
Exactly what I needed :D Thanks, I'll implement this because I also think the right way to go about it.

Reply:
Would it actually be more efficient to use a PartitionBy in this case? I believe it's going to be unbalanced as more studies end up using rally instead of the legacy stuff, and I don't understand the implications of choosing a particular fixed partition size as the decoder scales with volume.

Reply:
I'm referring to https://beam.apache.org/documentation/programming-guide/#partition

It's not doing anything more fancy than:

> Partition splits a single PCollection into a fixed number of smaller collections

So really, it just avoids scanning the collection twice to apply two different boolean predicates. But you end up having to do integer-based indexing which is ugly.

The filters probably read better. Once the filters are moved into `DecryptPayloads`, it might make sense to have the predicates be lambdas rather than named classes.

Reply:
Right: https://beam.apache.org/releases/javadoc/2.1.0/org/apache/beam/sdk/transforms/Partition.html

It does require specifying a numPartitions field, which is set for the window. I'm wondering if it has similar performance trade-offs as it does in Spark, where the number of partitions can limit the amount of parallelism e.g. 1 partition with 10 workers means 9 workers are left without work to do.

Reply:
I don't think that the same performance concepts exist here.

I believe the `numPartitions` field is necessary only for the SDK to know how many entries to provision in the output `PCollectionTuple`. The index of the tuple is significant, so every partition must be represented.

Reply:
Implemented. I liked the explicit fanout in the in the DAG, but performance first.